packages/ai-workspace-agent/src/browser/functions.ts

```diff
@@ -39,16 +53,65 @@ export class WorkspaceFunctionScope {
             throw new Error('Access outside of the workspace is not allowed');
         }
     }
-    /**
-     * Determines whether a given file or directory should be excluded from workspace operations.
-     *
-     * @param stat - The `FileStat` object representing the file or directory to check.
-     * @returns `true` if the file or directory should be excluded, `false` otherwise.
-     */
-    shouldExclude(stat: FileStat): boolean {
-        const excludedFolders = ['node_modules', 'lib'];
-        return stat.resource.path.base.startsWith('.') || excludedFolders.includes(stat.resource.path.base);
+
+    private async initializeGitignoreWatcher(workspaceRoot: URI): Promise<void> {
+        if (this.gitignoreWatcherInitialized) {
+            return;
+        }
+
+        const gitignoreUri = workspaceRoot.resolve(this.GITIGNORE_FILE_NAME);
+        this.fileService.watch(gitignoreUri);
+
+        this.fileService.onDidFilesChange(async event => {
+            if (event.contains(gitignoreUri)) {
+                this.gitignoreMatcher = undefined;
+            }
+        });
+
+        this.gitignoreWatcherInitialized = true;
+    }
+
+    async shouldExclude(stat: FileStat): Promise<boolean> {
+        const shouldConsiderGitIgnore = this.preferences.get(CONSIDER_GITIGNORE_PREF, false);
+        const userExcludePatterns = this.preferences.get<string[]>(USER_EXCLUDE_PATTERN_PREF, []);
+
+        if (this.isUserExcluded(stat.resource.path.base, userExcludePatterns)) {
+            return true;
+        }
+        const workspaceRoot = await this.getWorkspaceRoot();
+        if (shouldConsiderGitIgnore && await this.isGitIgnored(stat, workspaceRoot)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private isUserExcluded(fileName: string, userExcludePatterns: string[]): boolean {
+        return userExcludePatterns.some(pattern => new Minimatch(pattern, { dot: true }).match(fileName));
     }
+
+    private async isGitIgnored(stat: FileStat, workspaceRoot: URI): Promise<boolean> {
+        await this.initializeGitignoreWatcher(workspaceRoot);
+
+        const gitignoreUri = workspaceRoot.resolve(this.GITIGNORE_FILE_NAME);
+
+        try {
+            const fileStat = await this.fileService.resolve(gitignoreUri);
+            if (fileStat && !fileStat.isDirectory) {
+                if (!this.gitignoreMatcher) {
+                    const gitignoreContent = await this.fileService.read(gitignoreUri);
+                    this.gitignoreMatcher = ignore().add(gitignoreContent.value);
+                }
+                const relativePath = workspaceRoot.relative(stat.resource);
+                if (relativePath && this.gitignoreMatcher.ignores(relativePath.toString())) { return true; }
+            }
+        } catch {
+            // If .gitignore does not exist or cannot be read, continue without error
+        }
+
+        return false;
+    }
```

Comment:
```suggestion
    private async isGitIgnored(stat: FileStat, workspaceRoot: URI): Promise<boolean> {
        await this.initializeGitignoreWatcher(workspaceRoot);

        const gitignoreUri = workspaceRoot.resolve(this.GITIGNORE_FILE_NAME);

        try {
            const fileStat = await this.fileService.resolve(gitignoreUri);
            if (fileStat) {
                if (!this.gitignoreMatcher) {
                    const gitignoreContent = await this.fileService.read(gitignoreUri);
                    this.gitignoreMatcher = ignore().add(gitignoreContent.value);
                }
                const relativePath = workspaceRoot.relative(stat.resource);
                if (relativePath) {
                    const relativePathStr = relativePath.toString() + (stat.isDirectory ? '/' : '');
                    if (this.gitignoreMatcher.ignores(relativePathStr)) {
                        return true;
                    }
                }
            }
        } catch {
            // If .gitignore does not exist or cannot be read, continue without error
        }

        return false;
    }
```

I've debugged into the issue I had and understand now what's wrong:
1. we never filter directories at all based on the gitignore
2. the `ignoreMatcher` doesn't return `true` if in the gitignore it says `dist/` but the `relativePath` is `dist`

I adapted to code to imho work more as we would expect, but I'll leave it up to you if we want to filter directories. I think that we do, because otherwise the LLM would only see the directory but never a file it contains. This is ok for git (because it only deals with files anyway), but for the LLM it seems more correct to already remove the files.

Note that e.g. `dist` won't be filtered if the gitignore contains `dist/*`, but I'd argue with `dist/*` the user explicitly refers to contents of `dist` but not `dist` directly, while `dist/` is kind of the default everyone uses and it means, I want to ignore `dist`... but this is starting to get philosophical :-)